File: erlang/pe021.erl
-module(pe021).
-export([start/0]).

max() -> 10000.

start() ->
  pe021(max()).
  
pe021(N) ->
  S = pe021([], 1, 0, N),
  io:format("~w~n", [S]).

pe021(_L, N, S, M) when N =:= M ->
  S;

pe021(L, N, S, M) ->
  X = sum_factors(2, N, 1),
  if
    X < N ->
      I = lists:nth(N - X, L),
      if 
        I =:= N ->
          pe021([X | L], N + 1, X + N + S, M);
        true -> 
          pe021([X | L], N + 1, S, M)
      end;
    true ->
      pe021([X | L], N + 1, S, M)
  end.

sum_factors(I, N, S) when (I * I) > N ->
  S;

sum_factors(I, N, S) ->
  Q = N div I,
  R = N rem I,
  if
    R =:= 0 ->
      if
        I =:= Q ->
          sum_factors(I + 1, N, S + Q);
        true ->
          sum_factors(I + 1, N, S + Q + I)
      end;
    true ->
      sum_factors(I + 1, N, S)
  end.

File: ocaml/pe021.ml
open Core

let maxv = 10000 ;;

let rec sum_factors count num sum =
  if (count * count) > num then
    sum + 1
  else
    let quotient  = num / count in
    let remainder = num mod count in
    if remainder = 0 then
      if count = quotient then
        sum_factors (count + 1) num (sum + quotient)
      else
        sum_factors (count + 1) num (sum + quotient + count)
    else
      sum_factors (count + 1) num sum ;;

let rec pe021 list num sum maxv =
  if num = maxv then
    sum
  else
    let x = (sum_factors 2 num 0) in
    if x < num then
      let y = match List.nth list (num - x - 1) with
      | None -> 0
      | Some x -> x in
      if num = y then
        pe021 (x :: list) (num + 1) (x + num + sum) maxv
      else 
        pe021 (x :: list) (num + 1) sum maxv
    else
      pe021 (x :: list) (num + 1) sum maxv ;;

printf "%d\n" (pe021 [] 1 0 maxv) ;;
